AArch64: Add FMINNUM_IEEE and FMAXNUM_IEEE support

wzssyqa opened the pull request:
FMINNM/FMAXNM instructions of AArch64 follow IEEE754-2008. We can use them to canonicalize a floating point number. And FMINNUM_IEEE/FMAXNUM_IEEE is used by something like expanding FMINIMUMNUM/FMAXIMUMNUM, so let's define them.

Update combine_andor_with_cmps.ll.
Add fp-maximumnum-minimumnum.ll, with nnan testcases only.

V1F64 is not supported yet.
If we set v1f64 as legal, FMINNUM/FMAXNUM will have some problem:
   both of them use `if (isOperationLegalOrCustom(FMAXNUM_IEEE, VT))`.

AArch64 depends on `expandFMINNUM_FMAXNUM` returning `SDValue()`
for FMAXNUM and FMINNUM.

We should fix this problem, while it will be in future patch.

arsenm commented on llvm/test/CodeGen/AArch64/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,125 @@
+  %c = call nnan half @llvm.minimumnum.f16(half %a, half %b)
+  ret half %c
+}
+
```
Needs tests with illegal cases, like vectors that need splitting.

arsenm commented on llvm/test/CodeGen/AArch64/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,125 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=aarch64 --mattr=+fullfp16 < %s | FileCheck %s --check-prefix=AARCH64
+
+define <2 x double> @max_v2f64(<2 x double> %a, <2 x double> %b) {
```
All of these tests have nnan, but nnan is not mentioned in the name. need nnan and no-nnan variants of the tests

wzssyqa commented:
@arsenm ping.

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1345,18 +1350,30 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     }
 
     // AArch64 has implementations of a lot of rounding-like FP operations.
+    // And the same for FMAXNUM_IEEE and FMINNUM_IEEE.
```
Comment doesn't add anything

arsenm replied:
Yes, but that doesn't need clarification. The existing comment isn't really helpful either

wzssyqa replied:
OK. Let's remove them.

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1345,18 +1350,30 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
           setOperationAction(Op, Ty, Legal);
     }
 
+    // In fact TargetLowering::expandFMINIMUMNUM_FMAXIMUMNUM works well with
+    // them. While in narrowInsertExtractVectorBinOp, they are expected to be
+    // LegalOrCustom.
```
This comment is too specific and about details of other code. I'm not sure what "works well with them" means, just delete the whole thing?

arsenm replied:
This is just working around missing vector legalization support

wzssyqa replied:
Sorry. I don't understand it well.
The IR here is about 256bit operation, while NEON has 128bit support, thus a split is needed anyway.
Do you mean that we need to split it as an ISD operation?

arsenm replied:
The default vector legalization should give you the splitting result, this doesn't require custom lowering to handle

arsenm replied:
#109779 handles the widening case. You need to add the FMINIMUMNUM/FMAXIMUMNUM cases to DAGTypeLegalizer::SplitVectorResult, like is done for FMINIMUM/FMAXIMUM already

wzssyqa replied:
Coolll, thanks.

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -10229,6 +10249,28 @@ SDValue AArch64TargetLowering::LowerFCOPYSIGN(SDValue Op,
 }
 
+SDValue
+AArch64TargetLowering::LowerFMINIMUMNUM_FMAXIMUMNUM(SDValue Op,
```
This is repeating the generic expansion, this should not be necessary

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1345,18 +1350,30 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
+    // LegalOrCustom.
+    for (auto Op : {ISD::FMAXIMUMNUM, ISD::FMINIMUMNUM}) {
+      for (MVT Ty : {MVT::v4f32, MVT::v2f64})
+        setOperationAction(Op, Ty, Custom);
+      if (Subtarget->hasFullFP16())
+        setOperationAction(Op, MVT::v8f16, Custom);
```
These are custom and just repeat the generic expansion, this should not be necessary

wzssyqa committed:
AArch64: Add FMINNUM_IEEE and FMAXNUM_IEEE support

FMINNM/FMAXNM instructions of AArch64 follow IEEE754-2008.
We can use them to canonicalize a floating point number.
And FMINNUM_IEEE/FMAXNUM_IEEE is used by something like
expanding FMINIMUMNUM/FMAXIMUMNUM, so let's define them.

Update combine_andor_with_cmps.ll.
Add fp-maximumnum-minimumnum.ll, with nnan testcases only.

Known problem:
V1F64 is not supported yet.
If we set v1f64 as legal, FMINNUM/FMAXNUM will have some problem:
both of them use if (isOperationLegalOrCustom(FMAXNUM_IEEE, VT)).

AArch64 depends on expandFMINNUM_FMAXNUM returning SDValue()
for FMAXNUM and FMINNUM.

We should fix this problem, while it will be in future patch.

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.h:
```diff
@@ -1203,6 +1203,7 @@ class AArch64TargetLowering : public TargetLowering {
   SDValue LowerBitreverse(SDValue Op, SelectionDAG &DAG) const;
   SDValue LowerMinMax(SDValue Op, SelectionDAG &DAG) const;
   SDValue LowerFCOPYSIGN(SDValue Op, SelectionDAG &DAG) const;
+  SDValue LowerFMINIMUMNUM_FMAXIMUMNUM(SDValue Op, SelectionDAG &DAG) const;
```
Dead declaration

arsenm commented on llvm/test/CodeGen/AArch64/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,190 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=aarch64 --mattr=+fullfp16 < %s | FileCheck %s --check-prefix=AARCH64
+
+define <2 x double> @max_nnan_v2f64(<2 x double> %a, <2 x double> %b) {
```
Should test with and without nnan flag. Also should test some illegal vector widths, particularly 3 x and multiples of the legal vector sizes

wzssyqa replied:
For cases without nnan, we cannot add them now, since they depends on fcanonicalize.
I will add them with https://github.com/llvm/llvm-project/pull/104429.

wzssyqa replied:
For `multiples of the legal vector sizes`, I think that I have done them:
we have v4f64, v8f32, v16f16 here, and the length of NEON is 128bit.

arsenm commented on llvm/test/CodeGen/AArch64/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,190 @@
+; AARCH64-NEXT:    fminnm h0, h0, h1
+; AARCH64-NEXT:    ret
+entry:
+  %c = call nnan half @llvm.minimumnum.f16(half %a, half %b)
```
2 x half and 4 x half aren't tested, plus others

arsenm commented on llvm/test/CodeGen/AArch64/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,474 @@
+; AARCH64-NEXT:    // kill: def $d1 killed $d1 killed $q1
+; AARCH64-NEXT:    ret
+entry:
+  %c = call nnan <3 x double> @llvm.maximumnum(<3 x double> %a, <3 x double> %b)
```
non canonical suffix

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -5050,6 +5050,19 @@ def : Pat<(v1f64 (fmaxnum (v1f64 FPR64:$Rn), (v1f64 FPR64:$Rm))),
+          (FMINNMDrr FPR64:$a, FPR64:$b)>;
+def : Pat<(fminnum_ieee (f32 FPR32:$a), (f32 FPR32:$b)),
+          (FMINNMSrr FPR32:$a, FPR32:$b)>;
+def : Pat<(fminnum_ieee (f16 FPR16:$a), (f16 FPR16:$b)),
```
fp16 patterns require HasFullFP16 predicates, otherwise they could select instructions that are not present.

davemgreen reviewed (commented):
Thanks for doing this. Is the plan to change how fmax() is lowered? Otherwise this seems backwards.
